File: Day14/ScrollViewObjC/ScrollViewObjC/TapToZoom.m
//
//  TapToZoom.m
//  ScrollViewObjC
//
//  Created by techmaster on 8/21/14.
//  Copyright (c) 2014 Techmaster. All rights reserved.
//

#import "TapToZoom.h"
#define ZOOM_STEP 1.5

@interface TapToZoom () <UIScrollViewDelegate, UIGestureRecognizerDelegate>
//Ở đây chỉ cần khai báo weak, vì các control này sẽ được gắn lên ViewController.view
//ViewController.view sẽ có con trỏ strong giữ lại các property này
@property (weak, nonatomic) UIScrollView *scrollView;
@property (weak, nonatomic) UIImageView *photo;
@property (weak, nonatomic) UILabel* scaleLabel;
@end

@implementation TapToZoom

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.edgesForExtendedLayout = UIRectEdgeNone;
    //Configure scrollView
    //Chú ý: phải tạo local pointer trước, không gán trực tiếp đối tượng vào weak property
    UIScrollView* scrollView = [[UIScrollView alloc] initWithFrame:self.view.frame];
    scrollView.delegate = self;
    scrollView.minimumZoomScale = 0.1;
    scrollView.maximumZoomScale = 10;
    scrollView.zoomScale = 1.0;
    scrollView.clipsToBounds = YES;
    
    UIImageView* photo = [[UIImageView alloc] initWithFrame:scrollView.frame];
    photo.image = [UIImage imageNamed:@"playboy.jpg"];
    photo.contentMode = UIViewContentModeScaleAspectFit;
    //Phải có 2 lệnh dưới để hứng sự kiện người dùng tap vào photo
    photo.userInteractionEnabled = YES;
    photo.multipleTouchEnabled = YES;
    
    
    UITapGestureRecognizer* singleTap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(onTap:)];
    singleTap.numberOfTapsRequired = 1;
    singleTap.delegate = self;
    [photo addGestureRecognizer:singleTap];
    
    UITapGestureRecognizer* doubleTap = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(onDoubleTap:)];
    doubleTap.numberOfTapsRequired = 2;
    doubleTap.delegate = self;
    [photo addGestureRecognizer:doubleTap];
    
    //Nếu không có lệnh này, thì single tap sẽ nhận rồi sau đó double tap cũng nhận! Hình phóng ra rồi thu lại !
    [singleTap requireGestureRecognizerToFail:doubleTap];
    
    [scrollView addSubview:photo];
    [self.view addSubview:scrollView];
    
    //Sau khi scrollView, photo được gắn vào View cha, có con trỏ strong thì chúng
    //ta mới gán vào weak property
    self.scrollView = scrollView;
    self.photo = photo;
    
    UILabel* scaleLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, 0, 60, 40)];
    scaleLabel.textAlignment = NSTextAlignmentRight;
    scaleLabel.text = [NSString stringWithFormat:@"%2.2f", scrollView.zoomScale];
    
    UIBarButtonItem* barButton = [[UIBarButtonItem alloc] initWithCustomView:scaleLabel];
    
    [self.navigationItem setRightBarButtonItem:barButton];
    self.scaleLabel = scaleLabel;
}
- (UIView *)viewForZoomingInScrollView:(UIScrollView *)scrollView
{
    return self.photo;
}
- (void) onTap: (UITapGestureRecognizer*) tap {
    CGPoint tapPoint = [tap locationInView:self.photo];
    NSLog(@"x = %2.2f, y = %2.2f", tapPoint.x, tapPoint.y);
    
    float newScale = [self.scrollView zoomScale] * ZOOM_STEP;
    [self zoomRectForScale:newScale withCenter:[tap locationInView:self.photo]];
}

- (void) onDoubleTap: (UITapGestureRecognizer*) tap {
    
    float newScale = [self.scrollView zoomScale] / ZOOM_STEP;
    [self zoomRectForScale:newScale withCenter:[tap locationInView:self.photo]];

}
#pragma mark Utility methods

- (void)zoomRectForScale:(float)scale withCenter:(CGPoint)center {
    
    CGRect zoomRect;
    
    // the zoom rect is in the content view's coordinates.
    //    At a zoom scale of 1.0, it would be the size of the imageScrollView's bounds.
    //    As the zoom scale decreases, so more content is visible, the size of the rect grows.
    //Chú ý: nên tạo một local struct để lấy size tốc độ sẽ tốt hơn, không nên self.scrollView.bounds.size / scale
    
    CGSize scrollViewSize = self.scrollView.bounds.size;
    zoomRect.size.height = scrollViewSize.height / scale;
    zoomRect.size.width  = scrollViewSize.width  / scale;
    
    // choose an origin so as to get the right center.
    zoomRect.origin.x    = center.x - (zoomRect.size.width  / 2.0);
    zoomRect.origin.y    = center.y - (zoomRect.size.height / 2.0);
    
    [self.scrollView zoomToRect:zoomRect animated:YES];
    /*
     - (void)zoomToRect:(CGRect)rect animated:(BOOL)animated
     Parameters
     rect
     A rectangle defining an area of the content view. The rectangle should be in the coordinate space of the view returned by viewForZoomingInScrollView:
     ở đây viewForZoomingInScrollView trả về self.photo
     */
    self.scaleLabel.text = [NSString stringWithFormat:@"%2.2f", self.scrollView.zoomScale];
}
@end

File: Day14/ScrollInSwift/ScrollInSwift/TapToZoom.swift
//
//  TapToZoom.swift
//  ScrollInSwift
//
//  Created by Trung on 8/25/14.
//  Copyright (c) 2014 vn.Techmsater. All rights reserved.
//

import Foundation
import UIKit

let ZOOM_STEP:CGFloat = 2
class TapToZoom: UIViewController, UIScrollViewDelegate, UIGestureRecognizerDelegate {
    weak var scrollView: UIScrollView!
    weak var photo: UIImageView!
    weak var scaleLabel: UILabel!
    
    override func viewDidLoad() {
        // Chu y phai tao local pointer truoc, khong gan truc tiep doi tuong vao weak property
        var scrollView = UIScrollView(frame: self.view.frame)
        scrollView.delegate = self
        scrollView.minimumZoomScale = 0.1
        scrollView.maximumZoomScale = 10
        scrollView.clipsToBounds = true
        
        var photo = UIImageView(frame: scrollView.frame)
        photo.image = UIImage(named: "playboy.jpg")
        photo.contentMode = UIViewContentMode.ScaleAspectFit
        // Enabler user interaction
        photo.userInteractionEnabled = true
        photo.multipleTouchEnabled = true
        
        var singleTap = UITapGestureRecognizer(target: self, action: "onTap:")
        singleTap.numberOfTapsRequired = 1
        singleTap.delegate = self
        photo.addGestureRecognizer(singleTap)
        
        var doubleTap = UITapGestureRecognizer(target: self, action: "onDoubleTap:")
        doubleTap.numberOfTapsRequired = 2
        doubleTap.delegate = self
        photo.addGestureRecognizer(doubleTap)
        
        // Chu y den cau lenh duoi
        
        singleTap.requireGestureRecognizerToFail(doubleTap)
        
        scrollView.addSubview(photo)
        self.view.addSubview(scrollView)
        
        // Sau khi scrollView duoc gan vao view cha, co con tro strong thi ta moi gan vao weak property
        self.scrollView = scrollView
        self.photo = photo
        
        var scaleLabel = UILabel(frame: CGRectMake(0, 0, 60, 40))
        scaleLabel.textAlignment = NSTextAlignment.Right
        scaleLabel.text = NSString(format: "%2.2f", scrollView.zoomScale)
        
        var barbutton = UIBarButtonItem(customView: scaleLabel)
        self.navigationItem.setRightBarButtonItem(barbutton, animated: true)
        self.scaleLabel = scaleLabel
        
    }
    
    func viewForZoomingInScrollView(scrollView: UIScrollView!) -> UIView! {
        return self.photo
    }
    
    func onTap(sender: UITapGestureRecognizer) {
        var tapPoint = sender.locationInView(self.photo)
        var newScale = self.scrollView.zoomScale * ZOOM_STEP
        self.zoomRectForScale(newScale, center: tapPoint)
    }
    
    func onDoubleTap(sender: UITapGestureRecognizer){
        var newScale = self.scrollView.zoomScale / ZOOM_STEP
        self.zoomRectForScale(newScale, center: sender.locationInView(self.photo))
    }
    func zoomRectForScale(scale:CGFloat, center:CGPoint){
        var zoomRect: CGRect = CGRect()
        
        var scrollViewSize = self.scrollView.bounds.size
        zoomRect.size.height = scrollViewSize.height / scale
        zoomRect.size.width = scrollViewSize.width / scale
        
        zoomRect.origin.x = center.x - ( zoomRect.size.width / 2.0)
        zoomRect.origin.y = center.y - ( zoomRect.size.height / 2.0)
        
        self.scrollView.zoomToRect(zoomRect, animated: true)
        self.scaleLabel.text = "\(self.scrollView.zoomScale)"
    }
}